### ISSUE TITLE
glob misbehaviour regarding special characters and spaces

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8633]
Migrated from [rt.perl.org#40470](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40470) (status was 'resolved')

Searchable as RT40470$

**Comment by p5pRT** [comment]
### From dserrano5@yahoo.es

This is a bug report for perl from dserrano5@&#8203;yahoo\.es\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.




---
I've found what I think is a bug in File&#8203;::Glob when it comes to expand a
pattern that contains at least one \\s character\. Suppose we have two
directories\, one named 'foo\{barbaz' and other 'foo\\\{bar baz'\. Inside each one
of them there is a file named 'GOT IT\!'\. A typical expansion of all files
inside the directory without space could be&#8203;:

my $c = 'foo\\\{barbaz/\*';
print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
\_\_END\_\_
glob       \<foo\\\{barbaz/\*>
expands to \<foo\{barbaz/GOT IT\!>

If we try to do the same with the other directory\, we get&#8203;:

my $c = 'foo\\\{bar\\ baz/\*';
print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
\_\_END\_\_
glob       \<foo\\\{bar\\ baz/\*>
expands to \<foo>

We can achieve our purpose by adding extra backslashes&#8203;:

my $c = 'foo\\\\\\\{bar\\ baz/\*';
print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
\_\_END\_\_
glob       \<foo\\\\\{bar\\ baz/\*>
expands to \<foo\{bar baz/GOT IT\!>

But if we try the same with the first directory\, no expansion occurs&#8203;:

my $c = 'foo\\\\\\\{barbaz/\*';
print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
\_\_END\_\_
glob       \<foo\\\\\{barbaz/\*>
expands to \<foo\\>

This happens on these systems&#8203;:

1\. v5\.6\.1\, i386\-linux on Debian 3\.0
2\. v5\.8\.4\, ia64\-linux\-thread\-multi on Debian 3\.1
3\. v5\.8\.5\, i386\-linux\-thread\-multi on Red Hat Enterprise Linux ES rel\. 4
4\. v5\.8\.8\, ia64\-freebsd on FreeBSD 6\.1\-RELEASE
5\. v5\.8\.8\, i586\-linux\-thread\-multi on SUSE Linux Enterprise Server 10
6\. v5\.8\.8\, i486\-linux\-gnu\-thread\-multi on Debian 3\.1

Versions previous to 5\.6 behave in various ways but\, as they don't use
File&#8203;::Glob\, they are not relevant to this bug report\.

The relevant thread\[1\] that I started on PerlMonks narrowed the problem down
to how File&#8203;::Glob calls Text&#8203;::ParseWords&#8203;::parse\_line\. glob began to behave
correctly after I made this change \(also attached to this email\)&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- Glob.pm.orig        2006-10-08 02:02:10.000000000 +0200
+++ Glob.pm     2006-10-08 02:02:20.000000000 +0200
@@ -150,7 +150,7 @@
 	# implementation in Perl.  Need to support a flag
 	# to disable this behavior.
 	require Text::ParseWords;
-	@pat = Text::ParseWords::parse_line('\s+',0,$pat);
+	@pat = Text::ParseWords::parse_line('\s+',1,$pat);
     }
 
     # assume global context if not provided one


```
</details>
It applies cleanly to all systems enumerated above (with a little offset in
the case of perl 5\.6\)\, and makes glob work an all of them\. I also tried on
MSWin32 \(using ActiveState Perl versions 5\.6\.1\.629 and 5\.8\.8\.817\)\. The problem
exists\, but this patch does not fix it\. In fact\, it make things worse by
making glob useless when filenames have spaces\.

For your reference\, I'm including 32 snippets that shows the result in&#8203;:
 \- 5\.6\.1 and 5\.8\.8
 \- with original Glob\.pm and changed one
 \- Linux and Windows
 \- directory names with and without spaces
 \- patterns using one and three backslashes

Patch and test attached\. I'm not very good at writing tests so please bear
with me\.


References&#8203;:
 1\. http&#8203;://www\.perlmonks\.org/?node\_id=576225



Snippet index&#8203;:
ActiveState 5\.6\.1\.629 windows \(original\)\, filename without space
ActiveState 5\.6\.1\.629 windows \(original\)\, filename with space
ActiveState 5\.6\.1\.629 windows \(patched\)\, filename without space
ActiveState 5\.6\.1\.629 windows \(patched\)\, filename with space
5\.6\.1 linux \(original\)\, filename without space
5\.6\.1 linux \(original\)\, filename with space
5\.6\.1 linux \(patched\)\, filename without space
5\.6\.1 linux \(patched\)\, filename with space
ActiveState 5\.8\.8\.817 windows \(original\)\, filename without space
ActiveState 5\.8\.8\.817 windows \(original\)\, filename with space
ActiveState 5\.8\.8\.817 windows \(patched\)\, filename without space
ActiveState 5\.8\.8\.817 windows \(patched\)\, filename with space
5\.8\.8 linux \(original\)\, filename without space
5\.8\.8 linux \(original\)\, filename with space
5\.8\.8 linux \(patched\)\, filename without space
5\.8\.8 linux \(patched\)\, filename with space


ActiveState 5\.6\.1\.629 windows \(original\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

ActiveState 5\.6\.1\.629 windows \(original\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\{bar baz/GOT IT\!>

ActiveState 5\.6\.1\.629 windows \(patched\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

ActiveState 5\.6\.1\.629 windows \(patched\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\\>



5\.6\.1 linux \(original\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

5\.6\.1 linux \(original\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\{bar baz/GOT IT\!>

5\.6\.1 linux \(patched\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

5\.6\.1 linux \(patched\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\{bar baz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\\>



ActiveState 5\.8\.8\.817 windows \(original\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

ActiveState 5\.8\.8\.817 windows \(original\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\{bar baz/GOT IT\!>

ActiveState 5\.8\.8\.817 windows \(patched\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

ActiveState 5\.8\.8\.817 windows \(patched\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\\>



5\.8\.8 linux \(original\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

5\.8\.8 linux \(original\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\{bar baz/GOT IT\!>

5\.8\.8 linux \(patched\)\, filename without space
&nbsp;  my $c = 'foo\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{barbaz/\*>
&nbsp;  expands to \<foo\{barbaz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{barbaz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{barbaz/\*>
&nbsp;  expands to \<foo\\>

5\.8\.8 linux \(patched\)\, filename with space
&nbsp;  my $c = 'foo\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\{bar baz/GOT IT\!>

&nbsp;  my $c = 'foo\\\\\\\{bar\\ baz/\*';
&nbsp;  print "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">\\n";
&nbsp;  \_\_END\_\_
&nbsp;  glob       \<foo\\\\\{bar\\ baz/\*>
&nbsp;  expands to \<foo\\>




---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=medium


---
Site configuration information for perl v5\.8\.8&#8203;:

Configured by Debian Project at Sun Jun 11 16&#8203;:26&#8203;:10 UTC 2006\.

Summary of my perl5 \(revision 5 version 8 subversion 8\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.16\.4\, archname=i486\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux ninsei 2\.6\.16\.4 \#1 smp preempt tue apr 11 23&#8203;:28&#8203;:01 pdt 2006 i686 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-Dcccdlflags=\-fPIC \-Darchname=i486\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.8 \-Darchlib=/usr/lib/perl/5\.8 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.8\.8 \-Dsitearch=/usr/local/lib/perl/5\.8\.8 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Uusesfio \-Uusenm \-Duseshrplib \-Dlibperl=libperl\.so\.5\.8\.8 \-Dd\_dosuid \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.1\.2 20060604 \(prerelease\) \(Debian 4\.1\.1\-2\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=/lib/libc\-2\.3\.6\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.8\.8
&nbsp;    gnulibc\_version='2\.3\.6'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.8\.8&#8203;:
&nbsp;    /home/hue/lang/perl/modules/i486\-linux\-gnu\-thread\-multi
&nbsp;    /home/hue/lang/perl/modules
&nbsp;    /home/hue/lang/perl/modules/lib/perl/5\.8\.7
&nbsp;    /home/hue/lang/perl/modules/lib/perl/5\.8\.8
&nbsp;    /home/hue/lang/perl/modules/share/perl/5\.8\.7
&nbsp;    /home/hue/lang/perl/modules/share/perl/5\.8\.8
&nbsp;    /home/hue/lang/perl/modules/lib/i486\-linux\-gnu\-thread\-multi/
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.8\.8
&nbsp;    /usr/local/share/perl/5\.8\.8
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.8
&nbsp;    /usr/share/perl/5\.8
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    /usr/local/lib/perl/5\.8\.7
&nbsp;    \.



---
Environment for perl v5\.8\.8&#8203;:
&nbsp;    HOME=/home/hue
&nbsp;    LANG=es\_ES\.UTF\-8
&nbsp;    LANGUAGE=es\_ES\.UTF\-8
&nbsp;    LC\_ALL=es\_ES\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/hue/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/bin/X11&#8203;:/usr/games&#8203;:/home/hue/bin&#8203;:/home/hue/bin
&nbsp;    PERL5LIB=/home/hue/lang/perl/modules&#8203;:/home/hue/lang/perl/modules/lib/perl/5\.8\.7&#8203;:/home/hue/lang/perl/modules/lib/perl/5\.8\.8&#8203;:/home/hue/lang/perl/modules/share/perl/5\.8\.7&#8203;:/home/hue/lang/perl/modules/share/perl/5\.8\.8&#8203;:/home/hue/lang/perl/modules/lib/i486\-linux\-gnu\-thread\-multi/
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/bin/psh


\-\- 
 David Serrano

**Comment by p5pRT** [comment]
### From dserrano5@yahoo.es

<details><summary>Glob.pm.patch</summary>

```diff
--- /usr/lib/perl/5.8.8/File/Glob.pm.orig	2006-10-08 13:26:24.000000000 +0200
+++ /usr/lib/perl/5.8.8/File/Glob.pm	2006-10-08 12:16:29.000000000 +0200
@@ -150,7 +150,7 @@
 	# implementation in Perl.  Need to support a flag
 	# to disable this behavior.
 	require Text::ParseWords;
-	@pat = Text::ParseWords::parse_line('\s+',0,$pat);
+	@pat = Text::ParseWords::parse_line('\s+',1,$pat);
     }
 
     # assume global context if not provided one

```
</details>

**Comment by p5pRT** [comment]
### From dserrano5@yahoo.es

\#\!/usr/bin/perl

use strict;
use warnings;

my @&#8203;chars;
my $file = 'GOT IT\!';
BEGIN \{
&nbsp;    @&#8203;chars = grep \!m|^ \[/\\w\] $|x\, map chr\, 32\.\.126;
\}
use Test&#8203;::More tests => 4 \* @&#8203;chars;

foreach my $char \(@&#8203;chars\) \{
&nbsp;    my %dirs = \(
&nbsp;        nospace => \{
&nbsp;            proper => "foo$\{char\}barbaz"\,
&nbsp;            one\_bs => "foo\\\\$\{char\}barbaz"\,
&nbsp;            three\_bs => "foo\\\\\\\\\\\\$\{char\}barbaz"\,
&nbsp;        \}\,
&nbsp;        space => \{
&nbsp;            proper => "foo$\{char\}bar baz"\,
&nbsp;            one\_bs => "foo\\\\$\{char\}bar\\\\ baz"\,
&nbsp;            three\_bs => "foo\\\\\\\\\\\\$\{char\}bar\\\\ baz"\,
&nbsp;        \}\,
&nbsp;    \);

&nbsp;    foreach \(qw/nospace space/\) \{
&nbsp;        mkdir $dirs\{$\_\}\{proper\}\, 0755 or die "mkdir&#8203;: $\!";
&nbsp;        open my $fd\, '>'\, $dirs\{$\_\}\{proper\} \. '/' \. $file or die "open&#8203;: $\!";
&nbsp;        close $fd;
&nbsp;    \}

&nbsp;    is   \+\(glob $dirs\{nospace\}\{one\_bs\}   \. '/\*'\)\[0\]\, $dirs\{nospace\}\{proper\} \. '/' \. $file\, "glob on directory name containing '$char'\, without space";
&nbsp;    isnt \+\(glob $dirs\{nospace\}\{three\_bs\} \. '/\*'\)\[0\]\, $dirs\{nospace\}\{proper\} \. '/' \. $file\, "ditto\, using three backslashes";
&nbsp;    is   \+\(glob $dirs\{space\}\{one\_bs\}     \. '/\*'\)\[0\]\, $dirs\{space\}\{proper\}   \. '/' \. $file\, "glob on directory name containing '$char'\, with space";
&nbsp;    isnt \+\(glob $dirs\{space\}\{three\_bs\}   \. '/\*'\)\[0\]\, $dirs\{space\}\{proper\}   \. '/' \. $file\, "ditto\, using three backslashes";

&nbsp;    unlink $dirs\{nospace\}\{proper\} \. "/"\. $file\, $dirs\{space\}\{proper\} \. "/"\. $file;
&nbsp;    rmdir $dirs\{nospace\}\{proper\};
&nbsp;    rmdir $dirs\{space\}\{proper\};
\}

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Oct 08 04&#8203;:59&#8203;:21 2006\, dserrano5@&#8203;yahoo\.es wrote&#8203;:
> This is a bug report for perl from dserrano5@&#8203;yahoo\.es\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> I've found what I think is a bug in File&#8203;::Glob when it comes to expand
>    a
> pattern that contains at least one \\s character\. Suppose we have two
> directories\, one named 'foo\{barbaz' and other 'foo\\\{bar baz'\. Inside
>    each one
> of them there is a file named 'GOT IT\!'\. A typical expansion of all
>    files
> inside the directory without space could be&#8203;:

I can confirm that the original poster's observations persist in Perl
5\.14\.2 \(tested on Mac OS X 10\.4\.11\)&#8203;:

\#\!/usr/local/bin/perl
use strict; 
use warnings;
use feature qw\( :5\.14 \);

my $c = 'foo\\\{barbaz/\*';
say "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">";
say '';

my $d = 'foo\\\{bar\\ baz/\*';
say "glob       \<"\,$d\,">\\nexpands to \<"\, \(glob $d\)\, ">";
say '';

my $e = 'foo\\\\\\\{bar\\ baz/\*';
say "glob       \<"\,$e\,">\\nexpands to \<"\, \(glob $e\)\, ">";
say '';

my $f = 'foo\\\\\\\{barbaz/\*';
say "glob       \<"\,$f\,">\\nexpands to \<"\, \(glob $f\)\, ">";

Whether this is a problem that needs fixing is open to discussion\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Nov 06 10&#8203;:45&#8203;:41 2011\, jkeen@&#8203;verizon\.net wrote&#8203;:
> On Sun Oct 08 04&#8203;:59&#8203;:21 2006\, dserrano5@&#8203;yahoo\.es wrote&#8203;:
> > This is a bug report for perl from dserrano5@&#8203;yahoo\.es\,
> > generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > 
> > I've found what I think is a bug in File&#8203;::Glob when it comes to expand
> >    a
> > pattern that contains at least one \\s character\. Suppose we have two
> > directories\, one named 'foo\{barbaz' and other 'foo\\\{bar baz'\. Inside
> >    each one
> > of them there is a file named 'GOT IT\!'\. A typical expansion of all
> >    files
> > inside the directory without space could be&#8203;:
> 
> I can confirm that the original poster's observations persist in Perl
> 5\.14\.2 \(tested on Mac OS X 10\.4\.11\)&#8203;:
> 
> \#\!/usr/local/bin/perl
> use strict; 
> use warnings;
> use feature qw\( :5\.14 \);
> 
> my $c = 'foo\\\{barbaz/\*';
> say "glob       \<"\,$c\,">\\nexpands to \<"\, \(glob $c\)\, ">";
> say '';
> 
> my $d = 'foo\\\{bar\\ baz/\*';
> say "glob       \<"\,$d\,">\\nexpands to \<"\, \(glob $d\)\, ">";
> say '';
> 
> my $e = 'foo\\\\\\\{bar\\ baz/\*';
> say "glob       \<"\,$e\,">\\nexpands to \<"\, \(glob $e\)\, ">";
> say '';
> 
> my $f = 'foo\\\\\\\{barbaz/\*';
> say "glob       \<"\,$f\,">\\nexpands to \<"\, \(glob $f\)\, ">";
> 
> Whether this is a problem that needs fixing is open to discussion\.
> 
> Thank you very much\.
> Jim Keenan

I discovered this problem recently \(having absolutely no recollection of
this bug report\) and fixed it with commit 0b0e6d70\, but in doing so I
actually made things worse\, fixing them later in commit 1bb8785ab\.  I
didn’t actually realise until later how bad it had been between those
two commits \(various CPAN modules were failing their tests\)\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
